include/stl2/detail/memory/uninitialized_copy.hpp

```diff
@@ -0,0 +1,72 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Casey Carter 2016
+//  Copyright Christopher Di Bella 2016
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_DETAIL_ALGORITHM_UNINITIALIZED_COPY_HPP
+#define STL2_DETAIL_ALGORITHM_UNINITIALIZED_COPY_HPP
+
+#include <new>
+#include <stl2/iterator.hpp>
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/memory/destroy.hpp>
+#include <stl2/detail/memory/reference_to.hpp>
+#include <stl2/detail/tagged.hpp>
+
+STL2_OPEN_NAMESPACE {
+   ///////////////////////////////////////////////////////////////////////////
+   // uninitialized_copy [Extension]
+   //
+   template <InputIterator I, Sentinel<I> S, /*NoThrow*/ForwardIterator O>
+   requires
+      Constructible<value_type_t<O>, reference_t<I>>() &&
+      __ReferenceTo<O, value_type_t<O>>()
+   tagged_pair<tag::in(I), tag::out(O)>
+   uninitialized_copy(I first, S last, O result)
+   {
+      auto saved = result;
+      try {
+         for (; first != last; ++result, (void)++first) {
+            ::new(const_cast<void*>(static_cast<const volatile void*>(__stl2::addressof(*result))))
+               value_type_t<I>(*first);
```

Comment:
> `s/value_type_t<I>/value_type_t<O>/`

Proof my tests are not exhaustive :)

> Actually, I think all of the `::new`s would be better replaced with a helper function (somewhere under ) that deduces the proper type...
...so we don't have to keep reading that poor tortured placement new over and over.

We're standardising `destroy_at`; why not standardise `construct_at` too?